### ISSUE TITLE
Race conditions with __pycache__

### ISSUE THREAD

**Issue by abingham** [sixty-north/cosmic-ray#434]
There seem to be some potential race conditions with __pycache__ directories in our clones. There has been a verified report (separate issue coming soon, I think) where CR is definitely making the correct mutations on disk but the test process is still seeing an old (unmutated) version of the code. We have been able to make this problem go away by passing `-B` to the Python interpreter running the tests.

Perhaps this is because we're running too quickly, i.e. not letting the Python from one process "stop" before starting another one. Or perhaps it's an issue with multiprocessing. Who knows!

**Comment by abingham** [comment]
@LLKoder This is related to be bug you reported over email. I'm able to duplicate the problem you're seeing. I was able to work around it by adding `-B` to the `test-command` in your conf:
```
test-command = "python -B -m unittest discover test_calculator
```
Can you try this and let me know if it fixes things for you?

**Comment by abingham** [comment]
This is very interesting. If I run the tests in the original report multiple times, I get inconsistent results. Sometimes I get 34% survival, sometimes 45%, sometimes 54%. 

However, if I add a 1 second sleep after the `subprocess.run` for the test execution, I get consistent (and correct) results. Do we need to somehow ask the subprocess to "really, really finish"?

**Comment by abingham** [comment]
I've asked about this [on SO](https://stackoverflow.com/questions/56073827/why-does-python-ignore-source-files-in-favor-of-pyc-files-sometimes)

**Comment by abingham** [comment]
@LLKoder I've pushed a change that might fix the problem for you (d5c65a7ee2a569f088cfafe003662a504b354761). Can you pull from `master` and give it a try?

The basic issue as far as I can tell is that the changes CR writes to disk aren't always flushed do disk fast enough for the test-runner subprocesses to see them. So we end up in a situation where the tests one mutation are actually seeing an earlier mutation. So, for example, the replace-mul-with-div tests were seeing the mutation for replace-mul-with-add, resulting in an unexpected survivor.

Let me know if this changes works for you. I'm not 100% satisfied with it as a long-term solution, but it's not bad.

**Comment by LLKoder** [comment]
That's interesting for me too... :)
I haven't been still working with `subprocess`, but I know a little thread. Does `subprocess` have something like `thread.join()`

I tested:
test-command = "python -B -m unittest discover test_calculator
That's working...

I'll update CR branch and will test again.

**Comment by abingham** [comment]
> Does subprocess have something like thread.join()

Not really, but [the `multiprocessing` module](https://docs.python.org/3/library/multiprocessing.html) provides a threads-like API atop processes.

**Comment by LLKoder** [comment]
In report:
Fatal Python error: _Py_HashRandomization_Init: failed to get random numbers to initialize Python

All tests are killed. It's wrong. Must be two survived tests.

I have reseted `test-command`:
`test-command = "python -m unittest discover test_calculator`

Please add phase to the report: "Create time: mm dd yyyy HH:mm:ss",  if it's not hard. So I will see the difference between reports.

**Comment by abingham** [comment]
> In report:
> Fatal Python error: _Py_HashRandomization_Init: failed to get random numbers to initialize Python

This seems likely to be a problem with you Python setup. Can you try starting with a fresh virtual environment?

Also, what OS and python version are you using?

> Please add phase to the report: "Create time: mm dd yyyy HH:mm:ss", if it's not hard. So I will see the difference between reports.

What do you mean? You want a timestamp on each result in the database? And where would you like it displayed?

**Comment by LLKoder** [comment]
> This seems likely to be a problem with your Python setup. Can you try starting with a fresh virtual environment?
I use a fresh VE.

> Also, what OS and python version are you using?
Windows 10 64bit
Python 3.7.3 (v3.7.3:ef4ec6ed12, Mar 25 2019, 22:22:05) 64bit

> What do you mean? You want a timestamp on each result in the database? And where would you like it displayed?
I meant the date and time on a top of the report file.
Look at the screenshot.

![2019-05-10 22_32_25-Cosmic Ray Report](https://user-
images.githubusercontent.com/50442676/57528110-c16c1f80-7374-11e9-8ec3-48d815531f5e.png)

**Comment by abingham** [comment]
> I meant the date and time on a top of the report file.

That should be nearly trivial to implement. Care to take a try at it? The place you'll need to look is `src/cosmic_ray/tools/html.py`.

**Comment by abingham** [comment]
> I use a fresh VE.

Ok, then I'm not immediately sure what's going on. If I get some time I'll try this in a windows machine, but it may take a while...I've got work travel for the next few days.

**Comment by LLKoder** [comment]
I have commented [https://github.com/sixty-north/cosmic-ray/commit/d5c65a7ee2a569f088cfafe003662a504b354761]((d5c65a7) and return -B to `test-command`. Now all ok.

> That should be nearly trivial to implement. Care to take a try at it? The place you'll need to look is src/cosmic_ray/tools/html.py.
> (Also, I think it's a good idea.)

Ok, I'll try to do that tomorrow. I'm interesting, It will turn out or not. :) I'll send the script to you after the update, that it'll be in the repository.

**Comment by abingham** [comment]
The race condition is fixed as of d5c65a7ee2a569f088cfafe003662a504b354761. The solution (for now, at least) was to prohibit creation of pyc files in test runner subprocesses.

I'm going to deal with the `_Py_HashRandomization_init` problem in a different issue.
